pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/relocation/RealtimeSegmentRelocationManager.java

```diff
@@ -0,0 +1,287 @@
+    _runFrequencyMinutes = config.getRelocationManagerFrequencyInMinutes();
+
+    _executorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
+      @Override
+      public Thread newThread(Runnable runnable) {
+        Thread thread = new Thread(runnable);
+        thread.setName("PinotRelocationManagerExecutorService");
+        return thread;
+      }
+    });
+  }
+
+  public void start() {
+    LOGGER.info("Starting relocation manager");
+
+    _executorService.scheduleWithFixedDelay(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          runRelocation();
+        } catch (Exception e) {
+          LOGGER.warn("Caught exception while running relocation manager", e);
+        }
+      }
+    }, 5, _runFrequencyMinutes, TimeUnit.MINUTES);
+  }
+
+  public void stop() {
+    _executorService.shutdown();
+  }
+
+  /**
+   * Check all tables. Perform relocation of segments if table is realtime and relocation is required
+   */
+  public void runRelocation() {
+    if (!_pinotHelixResourceManager.isLeader()) {
+      LOGGER.info("Skipping realtime segment relocation, not leader!");
+      return;
+    }
+
+    LOGGER.info("Starting relocation of realtime segments");
+    List<String> allTableNames = _pinotHelixResourceManager.getAllTables();
+
+    for (String tableNameWithType : allTableNames) {
+
+      TableType tableType = TableNameBuilder.getTableTypeFromTableName(tableNameWithType);
+      if (tableType.equals(TableType.REALTIME)) {
+        LOGGER.info("Starting relocation of segments for table: {}", tableNameWithType);
+
+        TableConfig tableConfig = _pinotHelixResourceManager.getRealtimeTableConfig(tableNameWithType);
+
+        RealtimeTagConfig realtimeTagConfig = new RealtimeTagConfig(tableConfig, _helixManager);
+        if (!realtimeTagConfig.isRelocateCompletedSegments()) {
+          LOGGER.info("Skipping relocation of segments for {}", tableNameWithType);
+          return;
+        }
+
+        IdealState idealState = HelixHelper.getTableIdealState(_helixManager, tableNameWithType);
+        if (!idealState.isEnabled()) {
+          LOGGER.info("Skipping relocation of segments for {} since ideal state is disabled", tableNameWithType);
+          return;
+        }
+
+        Map<String, Map<String, String>> segmentNameToInstancesMap = relocateSegments(realtimeTagConfig, idealState);
+        if (!segmentNameToInstancesMap.isEmpty()) {
+          updateIdealState(tableNameWithType, segmentNameToInstancesMap);
+        }
+      }
+    }
+    LOGGER.info("Realtime segment relocation completed");
+  }
+
+  /**
+   * Given a realtime tag config and an ideal state, relocate the segments
+   * which are completed but still hanging around on consuming servers, one replica at a time
+   * @param  realtimeTagConfig
+   * @param idealState
+   */
+  protected Map<String, Map<String, String>> relocateSegments(RealtimeTagConfig realtimeTagConfig,
```

Comment:
Maybe the return type can be IdealState? (but the caller can use the map fields for replacing segments). We may need to document that it is a partial idealstate (containing only the segments that have been moved)